libs/util/data-fetcher/src/lib/data-fetcher.ts

```diff
@@ -64,68 +16,32 @@ class FetchError {
  *  2. otherwise, look for a Content-Type header in the response with a supported mime type
  *  3. if no valid mime type was found, look for an explicit file extension in the url (.csv, .geojson etc.)
  */
-export function readDataset(
+export async function readDataset(
   url: string,
   typeHint?: SupportedType
 ): Promise<DataItem[]> {
-  return useCache(
-    () =>
-      sharedFetch(url)
-        .catch((error) => {
-          throw FetchError.corsOrNetwork(error.message)
-        })
-        .then(async (response) => {
-          if (!response.ok) {
-            throw FetchError.http(response.status)
-          }
-          const fileInfo = parseHeaders(response.headers)
-          const fileExtensionMatches = new URL(
-            url,
-            typeof window !== 'undefined'
-              ? window.location.toString()
-              : undefined
-          ).pathname.match(/\.(.+)$/)
-          const fileExtension =
-            fileExtensionMatches && fileExtensionMatches.length
-              ? (fileExtensionMatches[1].toLowerCase() as SupportedType)
-              : null
-
-          let fileType: SupportedType
-
-          // 1. type hint
-          if (typeHint) fileType = typeHint
-          // 2. content-type header
-          else if ('supportedType' in fileInfo)
-            fileType = fileInfo.supportedType
-          // 3. file extension from url
-          else if (SupportedTypes.indexOf(fileExtension) > -1)
-            fileType = fileExtension
-
-          // no type inferred or hinted
-          if (!fileType) {
-            if ('mimeType' in fileInfo)
-              throw FetchError.unsupportedType(fileInfo.mimeType)
-            else throw FetchError.unknownType()
-          }
-
-          try {
-            switch (fileType) {
-              case 'csv':
-                return parseCsv(await response.text())
-              case 'json':
-                return parseJson(await response.text())
-              case 'geojson':
-                return parseGeojson(await response.text())
-              case 'excel':
-                return await parseExcel(await response.arrayBuffer())
-            }
-          } catch (e: any) {
-            throw FetchError.parsingFailed(e.message)
-          }
-          throw new Error('Not implemented')
-        }),
-    url
-  )
+  const fileType = await inferDatasetType(url, typeHint)
+
+  let dataset: BaseDataset
+  try {
+    switch (fileType) {
+      case 'csv':
+        dataset = new CsvDataset(url)
+        break
+      case 'json':
+        dataset = new JsonDataset(url)
+        break
+      case 'geojson':
+        dataset = new GeojsonDataset(url)
+        break
+      case 'excel':
```

Comment:
shouldn't you need a `default` as well ?

Reply:
all possible cases are covered here, if the file type cannot be guessed then an exception will be thrown before